File: src/main/native/cpp/Simulation/SimMotorManager/SimMotorManager.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "OvertureLib/Simulation/SimMotorManager/SimMotorManager.h"

#include "OvertureLib/MotorControllers/OverTalonFX/OverTalonFX.h"
#include <iostream>
#include <frc/RobotController.h>

SimMotorManager::SimMotorManager() {
	std::cout << "Created new Sim Motor Manager..." << std::endl;
}

void SimMotorManager::Init(
		const std::map<unsigned int, NTMotorName> CANIDToMotorNameMap) {
	std::cout << "Initializing SimMotorManager..." << std::endl;
	this->CANIDToMotorNameMap = CANIDToMotorNameMap;
	std::cout << "Got " << this->CANIDToMotorNameMap.size()
			<< " mapped motors..." << std::endl;

	std::cout << "Got " << this->motorsToRegister.size()
			<< " motors to register..." << std::endl;
	for (auto motor : this->motorsToRegister) {
		RegisterSimMotor(motor);
	}

	initialized = true;
}

void SimMotorManager::AddSimMotorCandidate(OverTalonFX *motor) {
	std::cout << "Adding sim motor candidate with id: " << motor->GetDeviceID()
			<< "..." << std::endl;
	this->motorsToRegister.emplace_back(motor);
}

void SimMotorManager::RegisterSimMotor(OverTalonFX *motor) {
	if (motor == NULL || motor == 0) {
		throw std::invalid_argument("SimMotorManager given null pointer!");
	}

	if (!this->CANIDToMotorNameMap.contains(motor->GetDeviceID())) {
		std::cout
				<< "SimMotorManager Warning: Tried to register a motor for simulation("
				<< motor->GetDeviceID() << ") that is not in the given map"
				<< std::endl;
		return;
	}

	std::string motorName = this->CANIDToMotorNameMap[motor->GetDeviceID()];

	if (this->registeredMotors.contains(motorName)) {
		std::cout
				<< "SimMotorManager Warning: Tried to register a motor for simulation that was already registered"
				<< std::endl;
		return;
	}

	std::shared_ptr < nt::NetworkTable > ntable = ntInst.GetTable(motorName);

	MotorNTPair newMotorPair;
	newMotorPair.motor = motor;
	newMotorPair.ntable = ntable;

	this->registeredMotors.emplace(motorName, std::move(newMotorPair));
}

void SimMotorManager::Update() {
	if (!initialized) {
		return;
	}
	for (auto motorIterator = this->registeredMotors.begin();
			motorIterator != this->registeredMotors.end(); motorIterator++) {
		MotorNTPair pair = motorIterator->second;

		std::shared_ptr < nt::NetworkTable > ntable = pair.ntable;
		OverTalonFX *motor = pair.motor;

		ctre::phoenix6::sim::TalonFXSimState &simState = motor->GetSimState();

		simState.SetSupplyVoltage(frc::RobotController::GetBatteryVoltage());

		bool inverted = motor->getCTREConfig().MotorOutput.Inverted.value;

		ntable->GetEntry("software_inverted").SetBoolean(inverted);

		units::volt_t motorVoltage = simState.GetMotorVoltage();
		ntable->GetEntry("voltage_applied").SetDouble(motorVoltage.value());

		units::turn_t motorPosition = units::turn_t(
				ntable->GetEntry("encoder_position").GetDouble(0));
		simState.SetRawRotorPosition(motorPosition);

		units::turns_per_second_t motorSpeed = units::turns_per_second_t(
				ntable->GetEntry("encoder_speed").GetDouble(0));
		simState.SetRotorVelocity(motorSpeed);
	}

	ntInst.Flush();
}

SimMotorManager& SimMotorManager::GetInstance() {
	static SimMotorManager instance;
	return instance;
}

File: src/main/java/com/overture/lib/simulation/SimMotorManager.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package com.overture.lib.simulation;

import com.ctre.phoenix6.signals.InvertedValue;
import com.ctre.phoenix6.sim.TalonFXSimState;
import com.overture.lib.motorcontrollers.OverTalonFX;
import edu.wpi.first.networktables.NetworkTable;
import edu.wpi.first.networktables.NetworkTableInstance;
import edu.wpi.first.wpilibj.RobotController;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/** Bridges every {@link OverTalonFX} to an external physics simulation over NetworkTables. */
public final class SimMotorManager {
  private static final SimMotorManager instance = new SimMotorManager();

  private final NetworkTableInstance ntInst = NetworkTableInstance.getDefault();

  private Map<Integer, String> canIdToMotorNameMap = new HashMap<>();
  private final Map<String, MotorNTPair> registeredMotors = new HashMap<>();
  private final List<OverTalonFX> motorsToRegister = new ArrayList<>();
  private boolean initialized = false;

  private static final class MotorNTPair {
    private NetworkTable ntable;
    private OverTalonFX motor;
  }

  private SimMotorManager() {
    System.out.println("Created new Sim Motor Manager...");
  }

  /**
   * Returns the singleton instance.
   *
   * @return the sim motor manager
   */
  public static SimMotorManager getInstance() {
    return instance;
  }

  /**
   * Queues a motor to be registered once {@link #init(Map)} runs.
   *
   * @param motor the motor to register
   */
  public void addSimMotorCandidate(OverTalonFX motor) {
    System.out.println("Adding sim motor candidate with id: " + motor.getDeviceID() + "...");
    motorsToRegister.add(motor);
  }

  /**
   * Registers every queued motor whose CAN id appears in the given map.
   *
   * @param canIdToMotorNameMap maps CAN id to the NetworkTables name of the simulated motor
   */
  public void init(Map<Integer, String> canIdToMotorNameMap) {
    System.out.println("Initializing SimMotorManager...");
    this.canIdToMotorNameMap = canIdToMotorNameMap;
    System.out.println("Got " + this.canIdToMotorNameMap.size() + " mapped motors...");

    System.out.println("Got " + motorsToRegister.size() + " motors to register...");
    for (OverTalonFX motor : motorsToRegister) {
      registerSimMotor(motor);
    }

    initialized = true;
  }

  private void registerSimMotor(OverTalonFX motor) {
    if (motor == null) {
      throw new IllegalArgumentException("SimMotorManager given null pointer!");
    }

    if (!canIdToMotorNameMap.containsKey(motor.getDeviceID())) {
      System.out.println(
          "SimMotorManager Warning: Tried to register a motor for simulation("
              + motor.getDeviceID()
              + ") that is not in the given map");
      return;
    }

    String motorName = canIdToMotorNameMap.get(motor.getDeviceID());

    if (registeredMotors.containsKey(motorName)) {
      System.out.println(
          "SimMotorManager Warning: Tried to register a motor for simulation that was already"
              + " registered");
      return;
    }

    MotorNTPair newMotorPair = new MotorNTPair();
    newMotorPair.motor = motor;
    newMotorPair.ntable = ntInst.getTable(motorName);

    registeredMotors.put(motorName, newMotorPair);
  }

  /** Exchanges motor state with the external simulation. Call this periodically. */
  public void update() {
    if (!initialized) {
      return;
    }

    for (MotorNTPair pair : registeredMotors.values()) {
      NetworkTable ntable = pair.ntable;
      OverTalonFX motor = pair.motor;

      TalonFXSimState simState = motor.getSimState();

      simState.setSupplyVoltage(RobotController.getBatteryVoltage());

      boolean inverted =
          motor.getCTREConfig().MotorOutput.Inverted == InvertedValue.Clockwise_Positive;

      ntable.getEntry("software_inverted").setBoolean(inverted);

      ntable.getEntry("voltage_applied").setDouble(simState.getMotorVoltage());

      simState.setRawRotorPosition(ntable.getEntry("encoder_position").getDouble(0));
      simState.setRotorVelocity(ntable.getEntry("encoder_speed").getDouble(0));
    }

    ntInst.flush();
  }
}
